File: app-service-comments/src/main/java/com/github/arhor/aws/graphql/federation/comments/service/impl/PostServiceImpl.java
package com.github.arhor.aws.graphql.federation.comments.service.impl;

import com.github.arhor.aws.graphql.federation.comments.data.entity.PostEntity;
import com.github.arhor.aws.graphql.federation.comments.data.repository.PostRepository;
import com.github.arhor.aws.graphql.federation.comments.generated.graphql.DgsConstants.POST;
import com.github.arhor.aws.graphql.federation.comments.generated.graphql.types.Post;
import com.github.arhor.aws.graphql.federation.comments.service.PostService;
import com.github.arhor.aws.graphql.federation.common.exception.EntityNotFoundException;
import com.github.arhor.aws.graphql.federation.common.exception.Operation;
import com.github.arhor.aws.graphql.federation.tracing.Trace;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.UUID;

@Trace
@Service
@RequiredArgsConstructor
public class PostServiceImpl implements PostService {

    private final PostRepository postRepository;

    @Override
    public Post findInternalPostRepresentation(final UUID postId) {
        return postRepository.findById(postId)
            .map(this::mapEntityToPost)
            .orElseThrow(() -> new EntityNotFoundException(
                POST.TYPE_NAME,
                POST.Id + " = " + postId,
                Operation.LOOKUP
            ));
    }

    @Override
    public void createInternalPostRepresentation(final UUID postId) {
        postRepository.save(
            PostEntity.builder()
                .id(postId)
                .build()
        );
    }

    @Override
    public void deleteInternalPostRepresentation(final UUID postId) {
        postRepository.deleteById(postId);
    }

    private Post mapEntityToPost(final PostEntity entity) {
        return Post.newBuilder()
            .id(entity.id())
            .build();
    }
}

File: dgs-comments/src/main/kotlin/com/github/arhor/dgs/comments/service/impl/PostServiceImpl.kt
package com.github.arhor.dgs.comments.service.impl

import com.github.arhor.dgs.comments.data.entity.PostEntity
import com.github.arhor.dgs.comments.data.repository.PostRepository
import com.github.arhor.dgs.comments.generated.graphql.types.Post
import com.github.arhor.dgs.comments.service.PostService
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class PostServiceImpl(
    private val postRepository: PostRepository,
) : PostService {

    override fun getPostById(postId: Long): Post? {
        return postRepository.findByIdOrNull(postId)?.let { Post(id = it.id) }
    }

    @Transactional
    override fun createPost(postId: Long) {
        postRepository.insert(PostEntity(id = postId))
    }

    @Transactional
    override fun deletePost(postId: Long) {
        postRepository.deleteById(postId)
    }
}
